docs/blueprints/EP038.rst

```diff
@@ -0,0 +1,68 @@
+:EP: 38
+:Title: Improving security to coloring application to switches
+:Authors:
+    - Aldo Ortega <aortegay AT fiu DOT edu>
+    - Italo Valcy <idasilva AT fiu DOT edu>
+
+:Created: 2023-06-27
+:Updated: 2023-07-20
+:Kytos-Version: 2023.2
+:Status: Draft
+
+*************************************************************
+EP38 - Improving security to coloring application to switches
+*************************************************************
+
+Abstract
+========
+
+Here it will be explained the vulnerability of **coloring** implementation related ot its broad match field. Also, it is going to be explained how these vulnerabilities could be solved.
+
+Motivation
+==========
+
+The NApp **coloring** installs flows according to a switch neighbors (NNI linked switches). These flows have three crucial characteristics:
+
+  1. The matching criteria has one field, **"dl_src"** which has the same standard as a **MAC** address; ie. **ee:ee:ee:ee:ee:02**.
+  2. The received packet is send to the controller.
+  3. These flows have the highest priority.
+
+Current implementation of **coloring** flows
+
+  .. image:: https://drive.google.com/uc?export=view&id=1G014OIYfxCuEYs_OEztaL9W9Rv80ZSWq
+
+**coloring** flows are meant to only accept packets from previously discovered NNIs but having only one match field **dl_src** does not assure it. Any connected interface with a matched **MAC** address can dangerously send loads of packets that could cause:
+
+  - Overload to the switch
+  - Overload to the controller
+  - Interruption to other flows of the affected switch
+
+Solution
+========
+
+We need to ensure that the packets that **coloring** is interested in, come from a discovered NNI and drop anything else. Adding the following should suffice:
+
+  - Adding **"in_port"** as matching criteria to the created flow. **coloring** recognizes only discovered NNIs. Their ports can be utilize to specify the matching so there is going to be one flow per NNI port. Examples flows 1 and 2.
+  - Add an extra flow to drop the unwanted traffic. It will match any packet with MAC prefix used by coloring. It can be done using mask. Example flow 3.
```

Comment:
The mask/wildcard idea is great to minimize the number of flows. Although currently, depending on a dpid value, for instance `cc:4e:24:4b:11:11:11:11`, [it can fill up the entire 6 bytes space without needing any of the 'ee' fill placeholder bytes](https://github.com/kytos-ng/coloring/blob/master/main.py#L130-L132), if we'll use mask:

- We'll need to reserve more bits just so we can be sure that wilcard will match accordingly
- We'll need to make sure and document which [`COLOR_FIELD`](https://github.com/kytos-ng/coloring/blob/master/settings.py#L3) will actually be supported and tested (only `dl_src` and `dl_dst` or should the solution also work for the other ones)?

Could you double check and cover this too?

Reply:
Checking if we could have a mask/wildcard that could cover all necessary cases. We are deleting `COLOR_FIELD` configurability so `dl_src` could just be a set value.

Reply:
Adding `dl_src` value to `settings.py` to have a default value `"ee:ee:ee:00:00:00/ff:ff:ff:00:00:00"`. It is enough to cover all current colors and makes it modifiable to be reduced or increased.

Reply:
@Alopalao right. The idea of exposing completely `DL_MASK` seems too much low level too me. Currently, we're not exposing the `b'ee'` byte filler. The idea of supporting a config here is great, maybe we needs to be exported is `dl_src_reserved_bytes = 3` (or something better named)?, meaning that the user will specify how many bytes are being reserved, then the first x MSB bytes would always be filled with `b'ee'` just so the wildcard would work as expected, and then at the runtime you compute the expected mask when sending sending the flows, what do you think? Let me know if you can see a safer way. Limiting the number of bytes also solves the problem of the dpid using the entire 6 byte space too.

Reply:
That looks great. We could name it `DL_MASK_BYTES`. But to confirm I am understanding well. For this variable to have:
- A value 5, means `dl_src = "ee:ee:ee:ee:ee:00/ff:ff:ff:ff:ff:00"
- A value 2, means `dl_src = "ee:ee:00:00:00:00/ff:ff:00:00:00:00"

Is this correct?

Reply:
Yes, @Alopalao. Great variable name `DL_MASK_BYTES` by the way. `6` - `DL_MASK_BYTES` represents how many bytes would be available for coloring each (dpid) neighbor, we need also a default value, probably 3 bytes (24 bits) is reasonable, especially since in the uplinks `in_port` will also be used. 

It's worth also confirming that indeed if a neighbor dpid that shares the same last 3 dpid bytes wouldn't cause issues (prior to this change each neighbor was guaranteed to have a different color since it was interpolating the entire 6 bytes). Maybe including the full example would be an opportunity to do so, unless you're crystal clear that it'll work then it's OK too.